Gems/Terrain/Code/Source/TerrainRenderer/TerrainFeatureProcessor.cpp

```diff
@@ -99,65 +100,146 @@ namespace Terrain
             AZ_Error(TerrainFPName, false, "Failed to create Terrain render buffers!");
             return;
         }
+        OnTerrainDataChanged(AZ::Aabb::CreateNull(), TerrainDataChangedMask::HeightData);
     }
 
     void TerrainFeatureProcessor::Deactivate()
     {
+        AzFramework::Terrain::TerrainDataNotificationBus::Handler::BusDisconnect();
+        AZ::RPI::MaterialReloadNotificationBus::Handler::BusDisconnect();
+
         m_patchModel = {};
         m_areaData = {};
-        AZ::RPI::MaterialReloadNotificationBus::Handler::BusDisconnect();
     }
 
     void TerrainFeatureProcessor::Render(const AZ::RPI::FeatureProcessor::RenderPacket& packet)
     {
         ProcessSurfaces(packet);
     }
 
-    void TerrainFeatureProcessor::UpdateTerrainData(
-        const AZ::Transform& transform,
-        const AZ::Aabb& worldBounds,
-        float sampleSpacing,
-        uint32_t width, uint32_t height, const AZStd::vector<float>& heightData)
+    void TerrainFeatureProcessor::OnTerrainDataDestroyBegin()
+    {
+        m_areaData = {};
+    }
+    
+    void TerrainFeatureProcessor::OnTerrainDataChanged(const AZ::Aabb& dirtyRegion, TerrainDataChangedMask dataChangedMask)
     {
-        if (!worldBounds.IsValid())
+        if (dataChangedMask != TerrainDataChangedMask::HeightData && dataChangedMask != TerrainDataChangedMask::Settings)
         {
             return;
         }
 
+        AZ::Aabb worldBounds = AZ::Aabb::CreateNull();
+        AzFramework::Terrain::TerrainDataRequestBus::BroadcastResult(
+            worldBounds, &AzFramework::Terrain::TerrainDataRequests::GetTerrainAabb);
+
+        const AZ::Aabb& regionToUpdate = dirtyRegion.IsValid() ? dirtyRegion : worldBounds;
+
+        m_dirtyRegion.AddAabb(regionToUpdate);
+        m_dirtyRegion.Clamp(worldBounds);
+
+        AZ::Transform transform = AZ::Transform::CreateTranslation(worldBounds.GetCenter());
+        
+        AZ::Vector2 queryResolution = AZ::Vector2(1.0f);
+        AzFramework::Terrain::TerrainDataRequestBus::BroadcastResult(
+            queryResolution, &AzFramework::Terrain::TerrainDataRequests::GetTerrainHeightQueryResolution);
+
+        uint32_t width = aznumeric_cast<uint32_t>((float)m_dirtyRegion.GetXExtent() / queryResolution.GetX());
+        uint32_t height = aznumeric_cast<uint32_t>((float)m_dirtyRegion.GetYExtent() / queryResolution.GetY());
+        
+        uint32_t worldWidth = aznumeric_cast<uint32_t>((float)worldBounds.GetXExtent() / queryResolution.GetX());
+        uint32_t worldHeight = aznumeric_cast<uint32_t>((float)worldBounds.GetYExtent() / queryResolution.GetY());
```

Comment:
Are the float casts necessary here?  
Also, nit - these could be const.

Reply:
floats aren't necessary, gonna move all this code straight to the areaData members.